### ISSUE TITLE
panel statusnotifier plugin should not block started apps and panel

### ISSUE THREAD

**Issue by agaida** [lxde/lxqt#708]
best viewed with quasselclient-qt5 - the client take ages to start, the panel is unusable until the application icon appears in the systemtray, a real no go

**Comment by pmattern** [comment]
I can confirm the finding regarding quasselclient.
But isn't it by any chance quasselclient itself which has to be blamed here? Just wondering as I couldn't see anything similar with any (of the very few) other applications using Plugin Status Notifier so far.

**Comment by agaida** [comment]
@pmattern - really don't know, but never seen this behaviour with xfce, windows, lxde, kde or gnome - all with notifications enabled - so it is a wild, but somewhat founded claim i think - maybe it is quassels fault, but i really don't know how to debug this.

**Comment by pmattern** [comment]
To my knowledge all desktop environments you state are providing *either* a plugin according to System Tray Protocol *or* a plugin implementing Status Notifier specification.
LXQt is the only one providing both, as @paulolieuthier just told me on IRC plugin Status notifier intentionally doesn't implement System Tray Protocol due to its poor quality besides a fallback functionality like this is actually part of the specification and the problem discussed in this issue seems to be related to using both LXQt plugins in parallel.
So maybe it's caused by that unique combination of features in LXQt.

Implementing Status Notifier in Qt is still wip, see e. g. https://wiki.qt.io/New_Features_in_Qt_5.5, section "QPA" (I'm pretty sure it was actually scheduled to be done in Qt 5.4 but can't find a source proving this right now).

All in all I think we should wait until Qt 5.5 is out before looking into this any further

**Comment by jleclanche** [comment]
The performance regression is a blocker for next release.

**Comment by stefonarch** [comment]
I don't know if it's related but I always noted this behaviour:
Adding the plugin akregator and konversation move from the systray there, removing it their icons disappear and don't move back to the systray, even restarting those apps; only restarting the panel shows the icons again normally.

**Comment by agaida** [comment]
some things i found out:
* https://bugreports.qt.io/browse/QTBUG-14485 - i have exactly the 25s blocking - so it might be on our side, eventually some of the remark could be helpful
* was about to file a bug in Quassel - until i noticed that we use org.kde...
<pre>
sendmsg(19, {msg_name(0)=NULL, msg_iov(2)=[{"l\1\0\1)\0\0\0]\0\0\0\220\0\0\0\1\1o\0\23\0\0\0/StatusN"..., 160}, {"\32\0\0\0org.kde.StatusNotifierItem\0\0"..., 41}], msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL) = 201
poll([{fd=19, events=POLLIN}], 1, 25000
</pre>

**Comment by palinek** [comment]
Can you, guys, check if the lxde/lxqt-panel#239 handles this?

**Comment by pmattern** [comment]
@palinek 
Had another look on up-to-date Arch Linux i686 and x86_64 in KVM/QEMU VMs with all LXQt components from a checkout this morning and lxqt-panel from branch notifier_fixes:
Basically the behaviour of Quassel remains exactly the same if both plugin-statusnotifier and plugin-tray are available. Some times the icon does come up in plugin-statusnotifier and then the problems addressed in this issue don't exist. Couldn't find out an eventual trigger that makes Quassel chose plugin-statusnotifier, though.
Given that the problem is apparently related to #817 I assume it's better to wait until the findings in [paulolieuthier's comment](https://github.com/lxde/lxqt/issues/817#issuecomment-142255137) are implemented before tracking this issue any further.

**Comment by paulolieuthier** [comment]
@pmattern please test with lxde/lxqt-qtplugin#5.

**Comment by pmattern** [comment]
Running lxde/lxqt-qtplugin@ca7f30a (branch `sni`) and lxde/lxqt-panel@2c293eb (branch `notifier_fixes`) on Arch i686 and x86_64 the problem is *not* fixed:
Quassel is now using plugin-statusnotifier by default as well. Most of the times it comes up at once and without issues. But sometimes it still comes up delayed and blocks the panel. The latter isn't reproducible 100% but seems to happen frequently when Quassel gets terminated and started againg within one LXQt session.

**Comment by paulolieuthier** [comment]
I just installed Quassel to help testing this. There are three packages for Quassel in Arch. quassel-server, quassel-client and quassel-monolithic. The last one doesn't not block, but quassel-client always block on startup. @pmattern what version of Quassel are you using?

**Comment by pmattern** [comment]
@paulolieuthier 
quassel-client. (Running on my test VM and connecting to Quassel core running on the host. FWIW the package providing Quassel core is `quassel-core`.)

**Comment by paulolieuthier** [comment]
> FWIW the package providing Quassel core is quassel-core.

Yeah, right. Please, test with quassel-monolithic.

**Comment by paulolieuthier** [comment]
I think I've found the reason why applications block, much related to what @palinek had said before in IRC. The reason is the absence of a context menu.

```cpp
#include <QApplication>
#include <QSystemTrayIcon>
#include <QMenu>

int main(int argc, char *argv[])
{
    QApplication a(argc, argv);

    QSystemTrayIcon icon(QIcon::fromTheme("folder-new"));

    QMenu menu("Title");
    menu.addAction(QIcon::fromTheme("document-save"), "Save");

    icon.setContextMenu(&menu); // comment this line and it'll block
    icon.show();

    return a.exec();
}
```

When with no menu, the application prints a message (actually, our plugin underneath it) before blocking:
```
QDBusConnection: error: could not send reply message to service "": Marshalling failed: Invalid object path passed in arguments
```

**Comment by pmattern** [comment]
Commits proposed in lxde/lxqt-panel#239 are fixing the problem with Quassel client (stand-alone / relying on Quassel core) without imparing Quassel monolithic.

Not 100% whether closing this issue which took place automatically due to the commit message was the right thing to do.
Eventually adopt a better solution from Plasma? Poke the Quassel folks about a possible bug in their code? @palinek, @paulolieuthier ?

**Comment by paulolieuthier** [comment]
@pmattern we read some Plasma's source code, but couldn't figure out what makes it different. We definetly should report the bug to Quassel's devs. I'll ask @palinek to do it tomorrow, as he can explain it better. The problem was that Quassel and lxqt-panel deadlock when trying to reach one another.

**Comment by palinek** [comment]
Not sure if we should reopen this one or open a new. I've made some testing and after all I think we should reimplement our logic to not block the GUI in any way.

IMO now an attacking application can make freeze lxqt-panel -> everything it needs is in cycle to preted to be a new StatusNotifierItem and don't respond to lxqt-panel queries (Title, Id...).

**Comment by paulolieuthier** [comment]
@pmattern the "better solution" you asked was implemented in lxde/lxqt#241.
